pkg/tunneler/integration_test.go

```diff
@@ -0,0 +1,228 @@
+
+	// private server
+	l, err := NewListener(publicServer.Client(), publicServer.URL, "d001")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// reverse proxy queries to an internal host
+	url, err := url.Parse(backend.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	proxy.Transport = backend.Client().Transport
+	server := &http.Server{Handler: proxy}
+	//nolint:errcheck
+	go server.Serve(l)
+
+	// client
+	// wait for the reverse connection to be established
+	time.Sleep(1 * time.Second)
+	stop := func() {
+		l.Close()
+		server.Close()
+		publicServer.Close()
+		backend.Close()
+	}
+	return publicServer.Client(), publicServer.URL + defaultTunnelPathPrefix + "/" + cmdTunnelProxy + "/d001/", stop
+
+}
+
+func Test_integration(t *testing.T) {
+	client, uri, stop := setup(t)
+	resp, err := client.Get(uri)
+	if err != nil {
+		t.Fatalf("Request Failed: %s", err)
+	}
+	defer resp.Body.Close()
+	defer stop()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		t.Fatalf("Reading body failed: %s", err)
+	}
+	// Log the request body
+	bodyString := string(body)
+	if bodyString != "Hello world" {
+		t.Errorf("Expected %s received %s", "Hello world", bodyString)
+	}
+}
+
+func Test_integration_multiple_connections(t *testing.T) {
+	client, uri, stop := setup(t)
+	defer stop()
+	var wg sync.WaitGroup
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resp, err := client.Get(uri)
+			if err != nil {
+				t.Errorf("Request Failed: %s", err)
+			}
+			defer resp.Body.Close()
+
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				t.Errorf("Reading body failed: %s", err)
+			}
+			// Log the request body
+			bodyString := string(body)
+			if bodyString != "Hello world" {
+				t.Errorf("Expected %s received %s", "Hello world", bodyString)
+			}
+		}()
+	}
+	wg.Wait()
+}
+
+func Test_integration_listener_reconnect(t *testing.T) {
+	backend := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Hello world")
+	}))
+	backend.EnableHTTP2 = true
+	backend.StartTLS()
+	defer backend.Close()
+
+	// public server
+	mux := http.NewServeMux()
+	apiHandler := WithSyncerTunnel(mux)
+	apiHandler = withClusterScope(apiHandler)
+	publicServer := httptest.NewUnstartedServer(apiHandler)
+	publicServer.EnableHTTP2 = true
+	publicServer.StartTLS()
+	defer publicServer.Close()
+
+	// private server
+	l, err := NewListener(publicServer.Client(), publicServer.URL, "d001")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// reverse proxy queries to an internal host
+	url, err := url.Parse(backend.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	proxy.Transport = backend.Client().Transport
+	server := &http.Server{Handler: proxy}
+	//nolint:errcheck
+	go server.Serve(l)
+	defer server.Close()
+
+	// client
+	// wait for the reverse connection to be established
+	time.Sleep(1 * time.Second)
```

Comment:
why 1s? Can this flake?

Reply:
it doesn't flake

```
./tunneler.test -test.run Test_integration_listener_reconnect
5s: 24 runs so far, 0 failures
10s: 48 runs so far, 0 failures
15s: 72 runs so far, 0 failures
20s: 96 runs so far, 0 failures
25s: 144 runs so far, 0 failures
30s: 168 runs so far, 0 failures
35s: 192 runs so far, 0 failures
40s: 216 runs so far, 0 failures
45s: 240 runs so far, 0 failures
50s: 288 runs so far, 0 failures
55s: 312 runs so far, 0 failures
1m0s: 336 runs so far, 0 failures
1m5s: 360 runs so far, 0 failures
```

Reply:
(now run it in a CPU-constrained environment)

Reply:
that is how we check the tests for Kubernetes/kubernetes , we didn't have flakes since we are doing that

Reply:
all credit is for Jordan Liggitt of course https://gist.github.com/liggitt/6a3a2217fa5f846b52519acfc0ffece0